src/main/java/org/truffleruby/core/thread/ThreadNodes.java

```diff
@@ -1035,4 +1044,71 @@ protected Object runBlockingSystemCall(Object executable, RubyArray argsArray,
             return foreignToRubyNode.executeConvert(result);
         }
     }
+
+    @Primitive(name = "thread_each_caller_location")
+    public abstract static class EachCallerLocationNode extends PrimitiveArrayArgumentsNode {
+
+        @Child private CallBlockNode yieldNode = CallBlockNode.create();
+
+        @Specialization
+        protected Object eachCallerLocation(VirtualFrame frame) {
+            final RubyProc block = (RubyProc) RubyArguments.getBlock(frame);
+            // Skip the block of `Thread#each_caller_location` + its internal iteration.
+            final int skip = 2;
+            final List<TruffleStackTraceElement> stackTraceElements = new ArrayList<>();
+            final AtomicInteger index = new AtomicInteger(0);
```

Comment:
Is the `AtomicInteger` concurrency primitive useful here?

Reply:
Don't know Java that well. To count skips in the iterator we need an outside value, which has to be `final`. Is there a better internally mutable form of int?
I was also recommended against implementing an anon visitor class for the iterator, which could have an added primitive int instance var.

Reply:
Got it.

But it's still not clear why the variable outside the iteration should be `final`. Could you please elaborate on it a bit?

Reply:
Because a Java lambda cannot set outer variables.

AtomicInteger is a lot of overhead, it's a full memory fence (which might involve sync'ing all CPUs) on every write.

> I was also recommended against implementing an anon visitor class for the iterator, which could have an added primitive int instance var.

This seems fine to me, and it is what is used in CallStackManager.
We should try to reuse `org.truffleruby.language.CallStackManager#iterateFrames` if possible

Reply:
Okay, I'll just expand it to an anon class and make a primitive int counter.

Reply:
Is it not possible to reuse CallStackManager#iterateFrames ? That would avoid duplication.

Reply:
Added: https://github.com/oracle/truffleruby/pull/3050/commits/aea27588b042e8be7e2853f5873750503c6b0ea8 I had to make CallStackManager#iterateFrames public, I assume that's fine.